packages/test/snapshots/src/sanitize.ts

```diff
@@ -3,124 +3,692 @@
+        return this.messages.length !== 0;
     }
-    return replaceTextCore(15);
 }
 
-function fixContents(messageContents: any) {
-    if (!messageContents ||
-        !messageContents.contents ||
-        !messageContents.contents.content ||
-        !messageContents.contents.content.contents) {
-        return;
+class Sanitizer {
+    readonly validator = new Validator.Validator();
+    // Represents the keys used to store Fluid object identifiers
+    readonly defaultExcludedKeys = new Set<string>();
+    // Represents the keys used by merge-tree ops their "seg" property, where other
+    // keys represent user information
+    readonly mergeTreeExcludedKeys = new Set<string>();
+    // Map of user information to what it was replaced with.  Used to ensure the same
+    // data have the same replacements
+    readonly replacementMap = new Map<string, string>();
+
+    /**
+     * Validate that the provided message matches the provided schema.
+     * For a full scrub, warn and continue (scrubber should fully sanitize unexpected
+     * fields for ops), otherwise throw an error because we cannot be sure user
+     * information is being sufficiently sanitized.
+     */
+    objectMatchesSchema = (object: any, schema: any): boolean => {
+        const result =  schema === false ? falseResult : this.validator.validate(object, schema);
+        if (!result.valid) {
+            const errorMsg = `Bad msg fmt:\n${result.toString()}\n${JSON.stringify(object, undefined, 2)}`;
+
+            if (this.fullScrub || this.noBail) {
+                console.error(errorMsg);
+            } else {
+                throw new Error(errorMsg);
+            }
+        }
+        return result.valid;
+    };
+
+    readonly chunkProcessor = new ChunkedOpProcessor(this.objectMatchesSchema);
+
+    constructor(
+        readonly messages: ISequencedDocumentMessage[],
+        readonly fullScrub: boolean,
+        readonly noBail: boolean,
+    ) {
+        this.defaultExcludedKeys.add("type");
+        this.defaultExcludedKeys.add("id");
+        this.defaultExcludedKeys.add("pkg");
+        this.mergeTreeExcludedKeys.add("nodeType");
+    }
+
+    isFluidObjectKey(key: string): boolean {
+        return key === "type" || key === "id";
     }
 
-    const contents = messageContents.contents.content.contents;
-    if (contents.seg) {
-        if (contents.seg.text) {
-            contents.seg.text = replaceText(contents.seg.text);
-        } else if (typeof contents.seg === "string") {
-            contents.seg = replaceText(contents.seg);
+    /**
+     * Replace text with garbage.  FluidObject types are not replaced when not under
+     * full scrub mode.  Non-Generic type text is replaced consistently.
+     */
+    replaceText(input?: string, type: TextType = TextType.Generic): string {
+        if (input === undefined) {
+            return undefined;
+        }
+
+        if (type === TextType.Generic) {
+            return "xxxxx";
```

Comment:
> return "xxxxx"; [](start = 11, length = 16)

this will break the merge tree. you need to replace with the same number of existing characters.

Reply:
in general i think you should always trying to replace strings with like strings. i would break things by word using regex, and map all words to the same value across everything you sanitize to attempt to preserve any hidden invariants as well.

---
In reply to: [500441391](https://github.com/microsoft/FluidFramework/pull/3786#discussion_r500441391) [](ancestors = 500441391)

Reply:
Changed all replacement to preserve length and for words to be replaced 1:1.  Clarified in the header that this reduces how "sanitized" the document actually is.  replay-tool reads the documents fine with these changes, but debugger tool doesn't.  debugger tool appears to have other issues/feature gaps that prevents this though, so I'll open other issues to track that.  I don't think it would be reasonable to cover debugger tool/fluid preview support in this PR.